File: cpp/functionOverriding/overriding.cpp
#include <iostream>
#include <cmath>
using namespace std;

class Shape
{
public:
    double getPerimeter()
    {
        return 0;
    }

    double getArea()
    {
        return 0;
    }

    double findExpenses()
    {
        return getArea() * 30;
    }
};

class Circle : public Shape
{
public:
    double radius;

    Circle(double r) : radius(r) {}

    double getPerimeter()
    {
        return 2 * M_PI * radius;
    }

    double getArea()
    {
        return M_PI * radius * radius;
    }

    double findExpenses()
    {
        return getArea() * 30;
    }
};

class Rectangle : public Shape
{
public:
    double length, breadth;

    Rectangle(double l, double b) : length(l), breadth(b) {}

    double getPerimeter()
    {
        return 2 * (length + breadth);
    }

    double getArea()
    {
        return (length * breadth);
    }

    double findExpenses()
    {
        return getArea() * 30;
    }
};

class Triangle : public Shape
{
public:
    double a, b, c;

    Triangle(double x, double y, double z) : a(x), b(y), c(z) {}

    double getPerimeter()
    {
        return a + b + c;
    }

    double getArea()
    {
        double s = getPerimeter() / 2;
        return sqrt(s * (s - a) * (s - b) * (s - c)); // Heron's formula
    }

    double findExpenses()
    {
        return getArea() * 30;
    }
};

int main()
{
    Circle circle(5);
    Rectangle rectangle(4, 6);
    Triangle triangle(3, 4, 5);

    cout << "Circle: \n";
    cout << "Perimeter: " << circle.getPerimeter() << " units\n";
    cout << "Area: " << circle.getArea() << " square units\n";
    cout << "Coloring Expense: Rs. " << circle.findExpenses() << "\n\n";

    cout << "Rectangle: \n";
    cout << "Perimeter: " << rectangle.getPerimeter() << " units\n";
    cout << "Area: " << rectangle.getArea() << " square units\n";
    cout << "Coloring Expense: Rs. " << rectangle.findExpenses() << "\n\n";

    cout << "Triangle: \n";
    cout << "Perimeter: " << triangle.getPerimeter() << " units\n";
    cout << "Area: " << triangle.getArea() << " square units\n";
    cout << "Coloring Expense: Rs. " << triangle.findExpenses() << "\n";

    return 0;
}

File: java/functionOverriding/Overriding.java
// package java.functionOverriding;
import java.util.Scanner;

class Shape {
    public double getArea() {
        return 0.0;
    }

    public double getPerimeter() {
        return 0.0;
    }

    public double getExpenses() {
        return getArea() * 30.0;
    }
}

class Circle extends Shape {
    double radius;

    Circle(double r) {
        radius = r;
    }

    public double getArea() {
        return (22.0 / 7.0) * radius * radius;
    }

    public double getPerimeter() {
        return 2.0 * (22.0 / 7.0) * radius;
    }
}

class Rectangle extends Shape {
    double length, breadth;

    Rectangle(double l, double b) {
        length = l;
        breadth = b;
    }

    public double getArea() {
        return length * breadth;
    }

    public double getPerimeter() {
        return 2 * (length + breadth);
    }
}

class Triangle extends Shape {
    double a, b, c;

    Triangle(double x, double y, double z) {
        a = x;
        b = y;
        c = z;
    }

    public double getArea() {
        double s = (a + b + c) / 2.0;
        double area = Math.sqrt(s * (s - a) * (s - b) * (s - c));
        return area;
    }

    public double getPerimeter() {
        return a + b + c;
    }
}

public class Overriding {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        double r;
        System.out.println("Enter Radius for Circle:");
        r = sc.nextDouble();
        Circle ci = new Circle(r);
        System.out.println("Area = " + ci.getArea());
        System.out.println("Perimeter = " + ci.getPerimeter());
        System.out.println("Expenses = " + ci.getExpenses() + "\n");

        double l, br;
        System.out.println("Enter Length & Breadth for Rectangle:");
        l = sc.nextDouble();
        br = sc.nextDouble();
        Rectangle rec = new Rectangle(l, br);
        System.out.println("Area = " + rec.getArea());
        System.out.println("Perimeter = " + rec.getPerimeter());
        System.out.println("Expenses = " + rec.getExpenses() + "\n");

        double a, b, c;
        System.out.println("Enter the dimensions of the Triangle:");
        a = sc.nextDouble();
        b = sc.nextDouble();
        c = sc.nextDouble();
        Triangle t = new Triangle(a, b, c);
        System.out.println("Area = " + t.getArea());
        System.out.println("Perimeter = " + t.getPerimeter());
        System.out.println("Expenses = " + t.getExpenses() + "\n");

        sc.close();
    }
}
